### ISSUE TITLE
project roadmap

### ISSUE THREAD

**Issue by mingfang** [joeycumines/go-behaviortree#1]
@joeycumines 
Thanks for giving us go-behaviortree. IMHO this is the best implementation.
Do you plan to continue with this project?
I have some enhancements and was wonder if I should fork or submit PR.

**Comment by joeycumines** [comment]
Hi @mingfang , thank you for the feedback!

I am always happy to review pull requests, it's something I quite enjoy. If you would like to make one please go ahead, I'll endeavor to give you useful feedback, at least. Fair warning though, I'll only merge changes that I feel align on matters like design, quality, and style, and only if I feel they deliver enough value in enough use cases to be included in the core implementation.

I'm quite curious to hear what enhancements you would like to make, could you elaborate a bit? The only truly critical components are the `Tick`, `Status` and `Node` types, and even they may be adapted to any equivalent implementation. Unless you are changing those core types (changing how trees are executed), even if you're changes are not merged, you should extend rather than fork this project, if that makes sense. Higher order functions are very powerful when working with the recursive `Node` type. It's also possible to plug in any sort of external ai / logic, though such logic ends up opaque / is ticked as a leaf node.

As far as a project roadmap goes, there is no planned or pending work. This project is definitely active, though, and I plan to continue to support it long term. The primary reason there is no planned work is simply because I consider the core functionality to be complete; I've solved all my immediate use cases quite convincingly. I have a couple of projects on the backburner that extend on this implementation, but they are involved enough that they will very likely stay that way unless another project develops a compelling need for either solution.

**Comment by mingfang** [comment]
@joeycumines Thanks for the response.

I'm trying to implement "memorized" nodes as describe in section `1.3.2 ControlFlow Nodes with Memory` in this research document https://arxiv.org/abs/1709.00084

My implementation looks like this
```go
package behaviortree

// Memorize wraps a Tick such that the Tick only runs once
func Memorize(tick Tick) Tick {
	if tick == nil {
		return nil
	}
	var (
		status Status
		err    error
		run    = false
	)
	return func(children []Node) (Status, error) {
		if !run {
			status, err = tick(children)
			run = true
		}
		return status, err
	}
}
```

**Comment by joeycumines** [comment]
That looks like a decent reference doc, thanks. I am definitely not well read on academic literature regarding behavior trees. I have just realised my first foray into them was just over 10 years ago now, which is a bit crazy.

I am finding it's meaning a tad opaque, without digging to find examples that use the `->*` notation, described in 1.3.2, but it appears to be describing the behavior of the existing `Sync` function. I do regret that name, in hindsight, since it's not particularly clear what it does. I implemented `Sync` immediately after the `Async` function. The need to avoid re-evaluating preceding child nodes became obvious as soon as I tried to utilise nodes that could return a running status. While I did not use the exact same mechanism in my implementation, doing `behaviortree.NewNode(behaviortree.Sequence, behaviortree.Sync(children))` should have the same effect, if my understanding is correct.

I have forgotten most of the context unfortunately, but I did come to the conclusion that `Tick` implementations making assumptions about their children beyond the duration of the execution they were provided them in was definitely not ideal. Part of my original intent, with the `Sync` implementation, was to facilitate conditional exit of running nodes, where the "memory" behavior needed to be applied across a subset of children of a given group tick, though I did then realise that would be possible with even at the node level (via nesting each group of children which needed different behavior). `Sync` has the benefit of working with any group tick that executes children in a stable order (between each non-running return), and returns running on encountering a running child.
